Fix pvpc_hourly_pricing by changing data source and modernise integration

azogue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

**Since 2021-11-29 the integration is broken** due to server-side banning (too much traffic for ~600 users at ~50 requests/day)

In #60795 a dep upgrade was proposed, which decreased the request rate to ~1 req/day, and also played cat-and-mouse with the `User-Agent` header to avoid the banning (so a not very polite quick solution for the problem), but that PR looks like it was forgotten in the pile 😅, @cgarwood had doubts about the headers trick, and I think I branched it from `master` instead of `dev` 🙈, and in top of that I've been absent for ~2 weeks without paying attention to  #60795 😰🙏

So this time I hope I'm doing it better 🥺

- This PR upgrades `aiopvpc` to v3.0.0, **changing the data source for the electricity prices to another official and documented public API** (from api.esios.ree.es to apidatos.ree.es 🤞)
- And simplifies a bit the integration and its test suite, by **upgrading to more modern** and standardized ways of doing cloud-polling API services 👍, including
  -  the **`DataUpdateCoordinator`** to handle data updates and availability
  - removing `RestoreEntity` inheritance for the `ElecPriceSensor`, as it has no meaning for this one.
  - using **`SensorEntityDescription`** and **`DeviceInfo`** attributes for a better definition of the price sensor, and get ready for future sensors for this service

The fixtures for the tests have been updated (and reduced), and, **if this PR is too big, it could be easily split in 2** (1st 3 commits as dep upgrade to **fix the broken integration**, and the next 4 for the 'modernize' part)

These changes don't affect the current documentation in https://www.home-assistant.io/integrations/pvpc_hourly_pricing/, so no need to change anything there 👌

#### Details for the changes in the 2 versions bumps for `aiopvpc`

- Changelog: https://github.com/azogue/aiopvpc/blob/master/CHANGELOG.md

- Release notes: https://github.com/azogue/aiopvpc/releases/tag/v3.0.0
- Diff/compare against current version: https://github.com/azogue/aiopvpc/compare/v2.3.0...v3.0.0

- Release notes: https://github.com/azogue/aiopvpc/releases/tag/v2.3.0
- Diff/compare against current version: https://github.com/azogue/aiopvpc/compare/v2.2.4...v2.3.0



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #60556
- This PR is related to issue: #60556, follow up #60795
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
:truck: Update test fixtures for pvpc_hourly_pricing

to switch to the new data source in `https://apidatos.ree.es`,
also using more recent dates, with the latest tariffs, for the tests.

azogue committed:
:white_check_mark: Adapt tests to new fixtures and dates

azogue committed:
:arrow_up: Update aiopvpc to change data source for electricity prices

to avoid recent banning from server using the current one. The new release uses
another public and documented API from the same organization, getting prices
from `apidatos.ree.es` instead of `api.esios.ree.es` 🤷

azogue committed:
:sparkles: Upgrade sensor platform to use DataUpdateCoordinator and modern descriptions

* Remove `RestoreEntity` inheritance for `ElecPriceSensor`,
  as it has no meaning for this one.
* Use `SensorEntityDescription` and `DeviceInfo` attributes for better
  definition of the price sensor, and get ready for future sensors for this service
* Add an `ElecPricesDataUpdateCoordinator` with the API handler,
  use `UpdateFailed` when price becomes unavailable, and relay on DataUpdateCoordinator
  for the availability state.

As the DUC can't manage the scheduled state changes at x:00h, AFAIK,
the `async_track_time_change` for the `update_current_price` is maintained in
the `async_added_to_hass`, but the call for download each half hour is now
handled by the DUC (the API handler won't call the external API each time,
it will only make 1-2 calls/day)

azogue committed:
:white_check_mark: Adjust sensor availability test to new behavior

azogue committed:
:pencil2: Fix lint warning

azogue committed:
:pencil2: Use SensorStateClass enum

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/__init__.py:
```diff
@@ -119,4 +129,43 @@ async def async_update_options(hass: HomeAssistant, entry: ConfigEntry) -> None:
+            if not prices:
+                raise UpdateFailed
+
+            self.api.process_state_and_attributes(now)
```
Can this call raise? Otherwise please move it out of the try... except block so we only wrap the line that can raise.

MartinHjelmare commented on homeassistant/components/pvpc_hourly_pricing/sensor.py:
```diff
@@ -1,160 +1,105 @@
-            self._pvpc_data.source_available = True
-            _LOGGER.warning("%s: component has recovered data access", self.entity_id)
-            self.update_current_price(now)
+        return self.coordinator.api.attributes
```
Please set the keys and values in a local dict explicitly before we return the dict as extra state attributes. We want to see what the keys are and their formatting and what the values mean.

azogue replied:
> We want to see what the keys are and their formatting and what the values mean.


The dict with lots of attrs is generated by the library since a long time, the logic was getting big and looked like the best approach to move out all of that and leaving this one lighter. 

Same for the test suite: logic for the download behavior, the DST changes, the tariff periods, and more little nightmares are deeply tested in `aiopvpc` 😅

Also, the keys for this dict are variable 😱. In the evening, the prices for the next day are available, and the dict has more info. In DST changes there is one more/one less hourly price tag, etc.

Logic is in https://github.com/azogue/aiopvpc/blob/600221a5cf82da4b96a7e47f4b8bc32a0fcad8b4/aiopvpc/pvpc_data.py#L245

I have in the TODO list to split some of those in multiple sensors, but didn't want to make breaking changes in this one, is that ok?

MartinHjelmare replied:
Please define a dictionary here in the integration with a map of state attribute keys and the keys that may be present in the data returned by the library. Then iterate over this map and set the items that are present in the state attributes.

azogue replied:
> Please define a dictionary here in the integration with a map of state attribute keys and the keys that may be present in the data returned by the library. Then iterate over this map and set the items that are present in the state attributes.

Hi @MartinHjelmare, that would be something like this 🙈😅:

```python
_KEYS_SENSOR_ATTRIBUTES = (
    "attribution",  # str
    # tariff data
    "tariff",  # str
    "period",  # str
    "available_power",  # int
    "next_period",  # str
    "hours_to_next_period",  # int
    # sensor stats
    "next_better_price",  # float
    "hours_to_better_price",  # int
    "num_better_prices_ahead",  # int
    "price_position",  # int
    "price_ratio",  # float
    "max_price",  # float
    "max_price_at",  # int
    "min_price",  # float
    "min_price_at",  # int
    "next_best_at",  # list[int]
    # hourly prices
    "price_00h",  # float
    "price_01h",  # float
    "price_02h",  # float
    "price_02h_d",  # float, only on DST day
    "price_03h",  # float
    "price_04h",  # float
    "price_05h",  # float
    "price_06h",  # float
    "price_07h",  # float
    "price_08h",  # float
    "price_09h",  # float
    "price_10h",  # float
    "price_11h",  # float
    "price_12h",  # float
    "price_13h",  # float
    "price_14h",  # float
    "price_15h",  # float
    "price_16h",  # float
    "price_17h",  # float
    "price_18h",  # float
    "price_19h",  # float
    "price_20h",  # float
    "price_21h",  # float
    "price_22h",  # float
    "price_23h",  # float
    # only at evening
    "next_better_price (next day)",  # float
    "hours_to_better_price (next day)",  # int
    "num_better_prices_ahead (next day)",  # int
    "price_position (next day)",  # int
    "price_ratio (next day)",  # float
    "max_price (next day)",  # float
    "max_price_at (next day)",  # int
    "min_price (next day)",  # float
    "min_price_at (next day)",  # int
    "next_best_at (next day)",  # list[int]
    "price_next_day_00h",  # float
    "price_next_day_01h",  # float
    "price_next_day_02h",  # float
    "price_next_day_02h_d",  # float, only on DST day
    "price_next_day_03h",  # float
    "price_next_day_04h",  # float
    "price_next_day_05h",  # float
    "price_next_day_06h",  # float
    "price_next_day_07h",  # float
    "price_next_day_08h",  # float
    "price_next_day_09h",  # float
    "price_next_day_10h",  # float
    "price_next_day_11h",  # float
    "price_next_day_12h",  # float
    "price_next_day_13h",  # float
    "price_next_day_14h",  # float
    "price_next_day_15h",  # float
    "price_next_day_16h",  # float
    "price_next_day_17h",  # float
    "price_next_day_18h",  # float
    "price_next_day_19h",  # float
    "price_next_day_20h",  # float
    "price_next_day_21h",  # float
    "price_next_day_22h",  # float
    "price_next_day_23h",  # float
)

self._attrs = {
    key: value
    for key, value in self.coordinator.api.attributes.items()
    if key in _KEYS_SENSOR_ATTRIBUTES
}
```

which doesn't has a lot of sense to my eyes, I don't know if I'm understanding and following you, sorry. The map is not required, as those `api.attributes` come with the desired keys, and the logic to include the keys is in the library, that's why, depending on the moment, the dict has _variable_ keys.

The typing is also kind of complex, but not problematic. The dict is typed as `dict[str, Any]` where the `Any` is really that `float | int | str | list[str]` shown in that `_KEYS_SENSOR_ATTRIBUTES` tuple of keys.

**How should I proceed?** Do you want the code above with the dict comprehension with that always-true if condition?

MartinHjelmare replied:
Yes, please. Also, we only want snake_case lowercase keys in the state attributes. So it would be good to make it a dict and normalise the keys in the state attributes like that.

The reason that we want this map is that the integration should be in control of what data is returned as state attributes.

Attribution is set as the corresponding entity property nowadays so we can move that item to that property.

MartinHjelmare replied:
Normalizing the keys can be done separately to avoid a breaking change.

azogue replied:
Ok, understood, thanks @MartinHjelmare  👍

> we only want snake_case lowercase keys in the state attributes

> Normalizing the keys can be done separately to avoid a breaking change

All keys are already normalized except the ones with the `"  (next day)"` suffix 🙈 that appear momentarily in the evening, and those are not in use in the custom lovelace card (https://github.com/danimart1991/pvpc-hourly-pricing-card), so I'll fix those with the new mapping (no need for another PR or even announce a breaking change for them, if you agree)

MartinHjelmare replied:
It's still a breaking change if they are changed. I think we should announce a breaking change to make it clear.

azogue replied:
@MartinHjelmare, a question:

The existence of the breaking change forces to include the PR in the next version `2022.1.0`? or it could be included in the next patch `2021.12.5`?

If the former, then I think I could do the key normalization in other PR, in order to deploy this one as soon as possible, as the integration is broken right now 🥺

MartinHjelmare replied:
Yes, do the breaking change in a separate PR. :+1:

azogue committed:
:fire: Remove extra fixtures and tests

as the `test_sensor_availability` is not needed anymore with the change to the
DataUpdateCoordinator. Now the same coverage can be reached with less 👍

azogue committed:
:art: Fix try-except clause

azogue committed:
:bug: Store sensor state and attributes

azogue committed:
:bug: Simplify fetch method for data update coordinator

No need to handle aiohttp.ClientError exception, already done in aiopvpc,
with asyncio.TimeoutError and others. `api.async_update_prices(now)` always
returns a dict, which may be empty.

azogue committed:
:art: Explicitly map the published sensor state attributes

without renaming them, except for the ones with `" (next day)" suffix, which are
normalized with the `"tomorrow_` prefix instead.

Also, store the `attribution` property in `_attr_attribution`

azogue committed:
:rewind: Revert rename of some sensor state attributes

to avoid breaking changes. Will do on next update.
